### PR TITLE
Optimization: duplicate balances

### PR DESCRIPTION
**Pull request by artur-abliazimov**

### CODE REVIEW

File: contracts/pool/src/methods/liquidate.rs [hunk 1]
```diff
@@ -152,20 +152,32 @@ fn do_liquidate(
...
+                who_collat_after = who_collat_after
+                    .checked_sub(s_token_to_burn)
+                    .ok_or(Error::MathOverflowError)?;
             }
 
-            if liquidator_collat_amount > 0 {
-                s_token.transfer_on_liquidation(who, liquidator, &liquidator_collat_amount);
+            if liquidator_collat_part > 0 {
+                let liquidator_collat_after = liquidator_collat_before
+                    .checked_add(liquidator_collat_part)
```
Comment (mn13) [inline contracts/pool/src/methods/liquidate.rs:170]: Is the `s_token_to_burn` substraction missing here?

File: contracts/pool/src/methods/withdraw.rs [hunk 2]
```diff
@@ -39,7 +40,7 @@ pub fn withdraw(
     let s_token = STokenClient::new(env, &reserve.s_token_address);
     let debt_token = DebtTokenClient::new(env, &reserve.debt_token_address);
```
Comment (mn13) [inline contracts/pool/src/methods/withdraw.rs:41]: Looks like debt token client used here only for getting it's address. Can we get address from reserve and avoid client instantiating?

File: contracts/pool/src/methods/repay.rs [hunk 3]
```diff
@@ -95,26 +93,36 @@ pub fn do_repay(
...
     let treasury_address = read_treasury(env);
 
     let underlying_asset = token::Client::new(env, asset);
+    let debt_token = DebtTokenClient::new(env, &reserve.debt_token_address);
+    let s_token = STokenClient::new(env, &reserve.s_token_address);
 
-    underlying_asset.transfer(who, &reserve.s_token_address, &lender_part);
-    underlying_asset.transfer(who, &treasury_address, &treasury_part);
+    underlying_asset.transfer(who, &reserve.s_token_address, &borrower_payback_amount);
+    s_token.transfer_underlying_to(&treasury_address, &treasury_part);
```
Comment (mn13) [inline contracts/pool/src/methods/repay.rs:120]: Wouldn't it be more profitable leave a underlying_asset.transfer here insted of s_token.transfer_underlying_to ?

File: contracts/pool/src/methods/withdraw.rs (hunk 2 above)
Comment (artur-abliazimov) [reply contracts/pool/src/methods/withdraw.rs:41]: Nice catch! Thanks!

File: contracts/pool/src/methods/repay.rs (hunk 3 above)
Comment (artur-abliazimov) [reply contracts/pool/src/methods/repay.rs:120]: Yes, it will be more efficient this way. Thanks!

**Review by mn13 — APPROVED** [review]

**Pull request merged**